nimbus/src/lib.rs

```diff
@@ -51,68 +47,102 @@ impl NimbusClient {
-            .map(|e| e.branch_slug.clone())
+            .map(|e| e.branch_slug.clone()))
     }
 
-    pub fn get_active_experiments(&self) -> Vec<EnrolledExperiment> {
-        self.enrolled_experiments.clone()
+    pub fn get_active_experiments(&self) -> Result<Vec<EnrolledExperiment>> {
+        get_enrollments(&self.db)
     }
 
-    pub fn get_all_experiments(&self) -> Vec<Experiment> {
-        self.experiments.clone()
+    pub fn get_all_experiments(&self) -> Result<Vec<Experiment>> {
+        self.db.collect_all(StoreId::Experiments)
     }
 
-    pub fn opt_in_with_branch(&self, _experiment_slug: String, _branch: String) {
-        unimplemented!()
+    pub fn opt_in_with_branch(&self, experiment_slug: String, branch: String) -> Result<()> {
+        opt_in_with_branch(&self.db, &experiment_slug, &branch)
     }
 
-    pub fn opt_out(&self, _experiment_slug: String) {
-        unimplemented!()
+    pub fn opt_out(&self, experiment_slug: String) -> Result<()> {
+        opt_out(&self.db, &experiment_slug)
     }
 
-    pub fn opt_out_all(&self) {
-        unimplemented!()
+    pub fn opt_out_all(&self) -> Result<()> {
+        for exp in self.get_all_experiments()? {
+            opt_out(&self.db, &exp.slug)?;
+        }
+        Ok(())
+    }
+
+    pub fn reset_enrollment(&self, experiment_slug: String) -> Result<()> {
+        reset_enrollment(
+            &self.db,
+            &experiment_slug,
+            &self.nimbus_id,
+            &self.available_randomization_units,
+        )
     }
 
     pub fn update_experiments(&self) -> Result<()> {
-        unimplemented!()
+        // I suspect we need to take some action when we find experiments we
+        // previously had no longer exist? For now though, just nuke them all.
+        log::info!("updating experiment list");
+        self.db.clear(StoreId::Experiments)?;
+        let experiments = self.http_client.get_experiments()?;
+        for experiment in experiments {
+            log::debug!("found experiment {}", experiment.slug);
+            self.db
+                .put(StoreId::Experiments, &experiment.slug, &experiment)?;
+        }
+        // Now update all enrollments based on the new set.
+        update_enrollments(
+            &self.db,
+            &self.nimbus_id,
+            &self.available_randomization_units,
+        )
     }
 
     pub fn nimbus_id(&self) -> Uuid {
         self.nimbus_id
     }
 
     fn get_or_create_nimbus_id(db: &Database) -> Result<Uuid> {
-        Ok(match db.get(DB_KEY_NIMBUS_ID)? {
+        Ok(match db.get(StoreId::Meta, DB_KEY_NIMBUS_ID)? {
             Some(nimbus_id) => nimbus_id,
             None => {
                 let nimbus_id = Uuid::new_v4();
-                db.put(DB_KEY_NIMBUS_ID, &nimbus_id)?;
+                db.put(StoreId::Meta, DB_KEY_NIMBUS_ID, &nimbus_id)?;
                 nimbus_id
             }
         })
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct EnrolledExperiment {
```

Comment:
Can we add something with the enrollment status to this, at least until we can record to Glean in Rust? That way the A-C component can record the enrollment status in telemetry.

Reply:
We can - but I'm not quite sure what you are asking for. The only reasons you can be enrolled are because you were selected, or you opted-in. Opting-in is a test only feature, so it's not clear it makes sense.

So maybe what you are really after is a list of experiments you were targeted for, but were not enrolled in? Or we rename this struct to something else and make some of the elements `Option<>` so it can hold both enrolled and "not selected" experiments? Let's chat more about this tomorrow.